modules/gapi/include/opencv2/gapi/rmat.hpp

```diff
@@ -0,0 +1,129 @@
+//
+// The device specific implementation is hidden in the RMat::Adapter class
+//
+// The basic flow is the following:
+// * Backend which is aware of the remote device:
+//   - Implements own AdapterT class which is derived from RMat::Adapter
+//   - Wraps device memory into RMat via make_rmat utility function:
+//         cv::RMat rmat = cv::make_rmat<AdapterT>(args);
+//
+// * End user:
+//   - Writes the code which works with RMats without any knowledge of the remote device:
+//     void func(const cv::RMat& in_rmat, cv::RMat& out_rmat) {
+//         // Fetch input data from the device, get mapped memory for output
+//         cv::RMat::View  in_view =  in_rmat.access(Access::R);
+//         cv::RMat::View out_view = out_rmat.access(Access::W);
+//         performCalculations(in_view, out_view);
+//         // data from out_view is transferred to the device when out_view is destroyed
+//     }
+class RMat
+{
+public:
+    // A lightweight wrapper on image data:
+    // - Doesn't own the memory;
+    // - Doesn't implement copy semantics (it's assumed that a view is created each time
+    // wrapped data is being accessed);
+    // - Has an optional callback which is called when the view is destroyed.
+    class View
+    {
+    public:
+        using DestroyCallback = std::function<void()>;
+
+        View() = default;
+        View(const GMatDesc& desc, uchar* data, size_t step = 0u, DestroyCallback cb = nullptr)
+            : m_desc(desc), m_data(data), m_step(step == 0u ? elemSize()*cols() : step), m_cb(cb)
+        {}
+
+        View(const View&) = delete;
+        View(View&&) = default;
+        View& operator=(const View&) = delete;
+        View& operator=(View&&) = default;
+        ~View() { if (m_cb) m_cb(); }
+
+        cv::Size size() const { return m_desc.size; }
+        const std::vector<int>& dims() const { return m_desc.dims; }
+        int cols() const { return m_desc.size.width; }
+        int rows() const { return m_desc.size.height; }
+        int type() const { return CV_MAKE_TYPE(depth(), chan()); }
+        int depth() const { return m_desc.depth; }
+        int chan() const { return m_desc.chan; }
+        size_t elemSize() const { return CV_ELEM_SIZE(type()); }
+
+        template<typename T = uchar> T* ptr(int y = 0, int x = 0) {
+            return reinterpret_cast<T*>(m_data + m_step*y + x*CV_ELEM_SIZE(type()));
+        }
+        template<typename T = uchar> const T* ptr(int y = 0, int x = 0) const {
+            return reinterpret_cast<const T*>(m_data + m_step*y + x*CV_ELEM_SIZE(type()));
+        }
+        size_t step() const { return m_step; }
+
+    private:
+        GMatDesc m_desc;
+        uchar* m_data = nullptr;
+        size_t m_step = 0u;
+        DestroyCallback m_cb = nullptr;
+    };
+
+    enum class Access { R, W };
+    class Adapter
+    {
+    public:
+        virtual ~Adapter() = default;
+        virtual GMatDesc desc() const = 0;
```

Comment:
Adapter can hold anything inside, it can be ie::blob for example, or some very specific memory description depending on the device. Putting `GMatDesc` here as a member looks like an unnecessary restriction. Yes, adapter implementor will need to create one to return in `desc()`, and implementor can add `GMatDesc` member if it will be useful, but let's leave this choice to implementor